src/client/debugger/jupyter/debuggingManager.ts

```diff
@@ -0,0 +1,213 @@
+        this.session = new Promise<vscode.DebugSession>((resolve, reject) => {
+            this.resolveFunc = resolve;
+            this.rejectFunc = reject;
+
+            vscode.debug
+                .startDebugging(undefined, {
+                    type: 'kernel',
+                    name: `${path.basename(document.uri.toString())}`,
+                    request: 'attach',
+                    internalConsoleOptions: 'neverOpen',
+                    __document: document.uri.toString()
+                })
+                .then(undefined, reject);
+        });
+    }
+
+    resolve(session: vscode.DebugSession) {
+        if (this.resolveFunc) {
+            this.resolveFunc(session);
+        }
+    }
+
+    reject(reason: Error) {
+        if (this.rejectFunc) {
+            this.rejectFunc(reason);
+        }
+    }
+
+    async stop() {
+        void vscode.debug.stopDebugging(await this.session);
+    }
+}
+
+/**
+ * The DebuggingManager maintains the mapping between notebook documents and debug sessions.
+ */
+@injectable()
+export class DebuggingManager implements IExtensionSingleActivationService {
+    private notebookToDebugger = new Map<vscode.NotebookDocument, Debugger>();
+    private readonly disposables: IDisposable[] = [];
+
+    public constructor(
+        @inject(IKernelProvider) private kernelProvider: IKernelProvider,
+        @inject(INotebookProvider) private notebookProvider: INotebookProvider,
+        @inject(IDebuggingCellMap) private debuggingCellMap: IDebuggingCellMap,
+        @inject(INotebookControllerManager) private readonly notebookControllerManager: INotebookControllerManager
+    ) {}
+
+    public async activate() {
+        vscode.debug.breakpoints; // start to fetch breakpoints
+
+        this.disposables.push(
+            // track termination of debug sessions
+            vscode.debug.onDidTerminateDebugSession(async (session) => {
+                for (const [doc, dbg] of this.notebookToDebugger.entries()) {
+                    if (dbg && session === (await dbg.session)) {
+                        this.debuggingCellMap.getCellsAnClearQueue(doc);
+                        this.notebookToDebugger.delete(doc);
+                        break;
+                    }
+                }
+            }),
+
+            // track closing of notebooks documents
+            vscode.workspace.onDidCloseNotebookDocument(async (document) => {
+                this.debuggingCellMap.getCellsAnClearQueue(document);
+                const dbg = this.notebookToDebugger.get(document);
+                if (dbg) {
+                    await dbg.stop();
+                }
+                this.fixBreakpoints(document);
+            }),
+
+            // factory for kernel debug adapters
+            vscode.debug.registerDebugAdapterDescriptorFactory('kernel', {
+                createDebugAdapterDescriptor: async (session) => {
+                    const activeDoc = vscode.window.activeNotebookEditor!.document;
```

Comment:
Using the ! here is breaking type safety, activeDoc should just be able to handle undefined or document cases.

Reply:
updated on the cleanup commit